build_tools/cmake/run_arm_sme_tests.py

```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env python3
+
+# Copyright 2024 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import sys
+import json
+import subprocess
+import os
+
+# This is a very simple test runner for tests that require Arm SME.
+# This assumes the host is an aarch64 machine.
+
+if __name__ == "__main__":
+    build_dir = "." if len(sys.argv) < 2 else sys.argv[1]
+    os.chdir(build_dir)
+
+    print("*************** Running ArmSME Tests ***************")
+
+    qemu_aarch64_executable = os.getenv("QEMU_AARCH64_EXECUTABLE", "qemu-aarch64")
+
+    # Gather list of tests that use ArmSME.
+    arm_sme_test_info = json.loads(
+        subprocess.check_output(
+            ["ctest", "--label-regex", "^requires-arm-sme$", "--show-only=json-v1"]
+        )
+    )
+
+    tests = arm_sme_test_info["tests"]
+
+    if len(tests) == 0:
+        print("No ArmSME tests found!", file=sys.stderr)
+        exit(1)
+
+    for test in tests:
+        name = test["name"]
+        test_command = [
+            qemu_aarch64_executable,
+            "-cpu",
+            "max,sme512=on,sme_fa64=off",
+            "--",
+            *test["command"],
+        ]
```

Comment:
> This script is a simple way to run the tests using a standard AArch64 build. The script is required as ctest (as far as I could find) does not allow simply injecting a prefix to the test commands without doing a full cross-compile (via CROSSCOMPILING_EMULATOR), which is overkill for just running a couple of SME tests.

Could these flags be passed via `RUNNER_ARGS` on `iree_generated_e2e_matmul_test`? Those get plumbed through to the test binary here: https://github.com/openxla/iree/blob/0d5bf27cb30cfa7f9a71bd7ff2818c8727c5f13e/build_tools/cmake/iree_native_test.cmake#L143-L148

Otherwise, we do have sections just above in that file for riscv and Android that wrap the test executable with other scripts. The `run_riscv_test` approach may be a good model to follow for this. I'd like to avoid having a separate test runner that looks through output from `ninja`, if we can avoid it.

Reply:
> Could these flags be passed via RUNNER_ARGS on iree_generated_e2e_matmul_test? 

The runner is `iree-e2e-matmul-test` not QEMU (there's nothing specific to QEMU about those tests, it's just a method to run them within the CI).

Reply:
> I'd like to avoid having a separate test runner that looks through output from ninja, if we can avoid it.

It's not looking at ninja output, it's just getting the list of tests ArmSME tests in JSON `ctest --label-regex ^requires-arm-sme$  --show-only=json-v1`.

Reply:
Sorry, I don't have a good model in my head for how some of these platforms are working. What should all this build system code bottom out at?

I see `Test command: ['qemu-aarch64', '-cpu', 'max,sme512=on,sme_fa64=off', '--', '/work/build-arm64/tools/iree-e2e-matmul-test', '--module=/work/build-arm64/tests/e2e/matmul/e2e_matmul_cpu_arm_sme_nondt_f32_small_llvm-cpu_local-task_sme_matmuls.vmfb', '--module=/work/build-arm64/tests/e2e/matmul/e2e_matmul_cpu_arm_sme_nondt_f32_small_llvm-cpu_local-task_sme_calls.vmfb', '--device=local-task']` in test logs: https://github.com/openxla/iree/actions/runs/8018957309/job/21905918125?pr=16331#step:6:94

How does that compare to a test _not_ using SME? Or to the other tests in the regular "Testing IREE" step: https://github.com/openxla/iree/actions/runs/8018957309/job/21905918125?pr=16331#step:5:15 ? (I don't want to special case things based on e2e_matmul, so seeing a job step just running those tests is concerning)

Reply:
There was nothing special about the e2e_matmul matmuls, that script simply ran tests tagged with `requires-arm-sme` under QEMU.

Anyway, I've changed things to insert the QEMU hooks in `iree_native_test.cmake`, which seemed to work locally. However, I can't tell looking at the CI logs if the ArmSME tests are running now. The e2e matmul tests 'pass' if:

1. They actually passed
2. or the CPU feature required to run them is missing

Both look the same in the logs, which is not great.

Reply:
> The e2e matmul tests 'pass' if:
> 
> 1. They actually passed
> 2. or the CPU feature required to run them is missing
> 
> Both look the same in the logs, which is not great.

Hmm... is that due to filtering at the CMake level? I made some changes to how the 'check' tests are configured that _define_ test but mark them as disabled if they are filtered. Could do the same for the matmul tests.

Reply:
Ah. Lovely. I had some similar soft failures in other tests and they led to unexpected behavior for developers months/years later :|